diff --git a/BUILD.gn b/BUILD.gn
index 2dbedac0..02e8b14e 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -212,6 +212,8 @@ static_library("libupdate_engine") {
     "cros/requisition_util.cc",
     "cros/shill_proxy.cc",
     "cros/update_attempter.cc",
+    "cros/jemaos_toggle_ota.cc",
     "libcurl_http_fetcher.cc",
     "metrics_utils.cc",
     "update_boot_flags_action.cc",


diff --git a/cros/jemaos_toggle_ota.cc b/cros/jemaos_toggle_ota.cc
new file mode 100644
index 00000000..c0397652
--- /dev/null
+++ b/cros/jemaos_toggle_ota.cc
@@ -0,0 +1,15 @@
+#include "update_engine/cros/jemaos_toggle_ota.h"
+#include <base/files/file_util.h>
+
+namespace jemaos {
+namespace {
+  // keep it the same with chromium src jemaos/misc/jemaos_toggle_ota.cc
+  const char kJemaOSOTAIndicatorFile[] = "/mnt/stateful_partition/unencrypted/preserve/.disable_jemaos_ota";
+}
+
+bool OTAEnabled() {
+  const bool enabled = !base::PathExists(base::FilePath(kJemaOSOTAIndicatorFile));
+  return enabled;
+}
+
+} // jemaos
diff --git a/cros/jemaos_toggle_ota.h b/cros/jemaos_toggle_ota.h
new file mode 100644
index 00000000..2c0fbb05
--- /dev/null
+++ b/cros/jemaos_toggle_ota.h
@@ -0,0 +1,8 @@
+#ifndef JEMAOS_TOGGLE_OTA_H_
+#define JEMAOS_TOGGLE_OTA_H_
+
+namespace jemaos {
+  bool OTAEnabled();
+}
+
+#endif // ifndef JEMAOS_TOGGLE_OTA_H_
diff --git a/cros/omaha_request_action.cc b/cros/omaha_request_action.cc
index a97408ba..d0a53805 100644
--- a/cros/omaha_request_action.cc
+++ b/cros/omaha_request_action.cc
@@ -55,6 +55,9 @@
 #include "update_engine/cros/payload_state_interface.h"
 #include "update_engine/cros/update_attempter.h"
 #include "update_engine/metrics_utils.h"
+// ---***JEMAOS BEGIN***---
+#include "update_engine/cros/jemaos_toggle_ota.h"
+// ---***JEMAOS END***---
 
 using base::Time;
 using base::TimeDelta;
@@ -1300,6 +1303,11 @@ bool OmahaRequestAction::ShouldIgnoreUpdate(ErrorCode* error) const {
     LOG(ERROR) << "All packages were excluded.";
   }
 
+  if (!jemaos::OTAEnabled()) {
+    LOG(INFO) << "jemaos ota disabled, ignore update";
+    return true;
+  }
+
   // Note: We could technically delete the UpdateFirstSeenAt state when we
   // return true. If we do, it'll mean a device has to restart the
   // UpdateFirstSeenAt and thus help scattering take effect when the AU is
diff --git a/cros/omaha_request_builder_xml.cc b/cros/omaha_request_builder_xml.cc
index 9eeb8302..291878cb 100644
--- a/cros/omaha_request_builder_xml.cc
+++ b/cros/omaha_request_builder_xml.cc
@@ -36,6 +36,9 @@
 #include "update_engine/common/utils.h"
 #include "update_engine/cros/omaha_request_params.h"
 #include "update_engine/cros/update_attempter.h"

 
 using std::string;
 
@@ -302,6 +305,11 @@ string OmahaRequestBuilderXml::GetApp(const OmahaAppData& app_data) const {
     app_versions = "version=\"" +
                    XmlEncodeWithDefault(app_data.version, kNoVersion) + "\" ";
   }

   string download_channel = params->download_channel();
   string app_channels =
@@ -410,6 +418,9 @@ string OmahaRequestBuilderXml::GetApp(const OmahaAppData& app_data) const {
       // DLC excluded for installs and updates.
       (app_data.is_dlc ? "" : requisition_arg) +
 

       ">\n" +
          app_body +
       "    </app>\n";
diff --git a/cros/update_attempter.cc b/cros/update_attempter.cc
index 184f9453..3a0b0330 100644
--- a/cros/update_attempter.cc
+++ b/cros/update_attempter.cc
@@ -71,6 +71,7 @@
 #include "update_engine/update_manager/omaha_request_params_policy.h"
 #include "update_engine/update_manager/update_manager.h"
 #include "update_engine/update_status_utils.h"
+#include "update_engine/cros/jemaos_toggle_ota.h"
 
 using base::FilePath;
 using base::Time;
@@ -208,6 +209,11 @@ bool UpdateAttempter::ScheduleUpdates(const ScheduleUpdatesParams& params) {
     return false;
   }
 
+  if (!jemaos::OTAEnabled()) {
+    LOG(INFO) << "jemaos ota disabled, skip scheduling updates";
+    return false;
+  }
+
   // We limit the async policy request to a reasonably short time, to avoid a
   // starvation due to a transient bug.
   policy_data_.reset(new UpdateCheckAllowedPolicyData());
@@ -1044,6 +1050,13 @@ bool UpdateAttempter::CheckForUpdate(
     LOG(INFO) << "Update check is only going to query server for update, will "
               << "not be applying any updates.";
   }
+  // ---***JEMAOS BEGIN***---
+  if (!jemaos::OTAEnabled()) {
+    LOG(INFO) << "jemaos ota disabled, refuse to check for update";
+    BroadcastStatus();
+    return false;
+  }
+  // ---***JEMAOS END***---
 
   LOG(INFO) << "Forced update check requested.";
   forced_app_version_.clear();
diff --git a/update_manager/official_build_check_policy_impl.cc b/update_manager/official_build_check_policy_impl.cc
index e1c55e31..263f72a9 100644
--- a/update_manager/official_build_check_policy_impl.cc
+++ b/update_manager/official_build_check_policy_impl.cc
@@ -26,6 +26,8 @@ EvalStatus OnlyUpdateOfficialBuildsPolicyImpl::Evaluate(
     State* state,
     std::string* error,
     PolicyDataInterface* data) const {
+// ---***JEMAOS BEGIN***---
+  /*
   const bool* is_official_build_p =
       ec->GetValue(state->system_provider()->var_is_official_build());
   if (is_official_build_p != nullptr && !(*is_official_build_p)) {
@@ -40,6 +42,8 @@ EvalStatus OnlyUpdateOfficialBuildsPolicyImpl::Evaluate(
     LOG(INFO) << "Unofficial build, but periodic update check interval "
               << "timeout is defined, so update is not blocked.";
   }
+  */
+// ---***JEMAOS END***---
   return EvalStatus::kContinue;
 }
 
